sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformation.scala

```diff
@@ -53,28 +82,176 @@ case class ScriptTransformation(
       val inputStream = proc.getInputStream
       val outputStream = proc.getOutputStream
       val reader = new BufferedReader(new InputStreamReader(inputStream))
+ 
+      val outputSerde: AbstractSerDe = if (outputSerdeClass != "") {
+        val trimed_class = outputSerdeClass.split("'")(1) 
+        Utils.classForName(trimed_class)
+          .newInstance.asInstanceOf[AbstractSerDe]
+      } else {
+        null
+      }
+ 
+      if (outputSerde != null) {
+        val columns = output.map { case aref: AttributeReference => aref.name }
+          .mkString(",")
+        val columnTypes = output.map { case aref: AttributeReference =>
+          aref.dataType.toTypeInfo.getTypeName()
+        }.mkString(",")
+
+        var propsMap = outputSerdeProps.map(kv => {
+          (kv._1.split("'")(1), kv._2.split("'")(1))
+        }).toMap + (serdeConstants.LIST_COLUMNS -> columns)
+        propsMap = propsMap + (serdeConstants.LIST_COLUMN_TYPES -> columnTypes)
+
+        val properties = new Properties()
+        properties.putAll(propsMap)
+ 
+        outputSerde.initialize(null, properties)
+      }
+
+      val outputSoi = if (outputSerde != null) {
+        outputSerde.getObjectInspector().asInstanceOf[StructObjectInspector]
+      } else {
+        null
+      }
+
+      val iterator: Iterator[Row] = new Iterator[Row] with HiveInspectors {
+        var cacheRow: Row = null
+        var curLine: String = null
+        var eof: Boolean = false
+
+        override def hasNext: Boolean = {
+          if (outputSerde == null) {
+            if (curLine == null) {
+              curLine = reader.readLine()
+              curLine != null
+            } else {
+              true
+            }
+          } else {
+            !eof
+          }
+        }
 
-      // TODO: This should be exposed as an iterator instead of reading in all the data at once.
-      val outputLines = collection.mutable.ArrayBuffer[Row]()
-      val readerThread = new Thread("Transform OutputReader") {
```

Comment:
If I am wrong please let me know, but I suppose that because I use an iterator here so it shouldn't already be the streaming style output and do as what the `TODO` wants?